internal/states/encryption/encryption.go

```diff
@@ -0,0 +1,271 @@
+package encryption
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionflow"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+)
+
+// New creates a new Encryption object. You can use this object to feed in encryption configuration and then create
+// an encryptionflow.StateFlow or encryptionflow.PlanFlow as needed.
+//
+// Note: a large portion of the OpenTofu codebase is still procedural, which means there is no way to properly inject
+// the Encryption object and carry the information it holds across subsystem boundaries. In most cases you should use
+// GetSingleton() to get a globally scoped copy of the Encryption object. However, for tests you should use this
+// function and hopefully, some time in the future, we can get rid of the singleton entirely.
+func New(logger hclog.Logger) Encryption {
+	return &encryption{
+		encryptionConfigs:         make(encryptionconfig.ConfigMap),
+		decryptionFallbackConfigs: make(encryptionconfig.ConfigMap),
+		mutex:                     sync.Mutex{},
+		logger:                    logger,
+	}
+}
+
+// Encryption is the main interface to feed the encryption configuration and obtain the encryptionflow.Flow for running
+// the actual encryption.
+//
+// Note: a large portion of the OpenTofu codebase is still procedural, which means there is no way to properly inject
+// the Encryption object and carry the information it holds across subsystem boundaries. In most cases you should use
+// GetSingleton() to get a globally scoped copy of the Encryption object. However, for tests you should use the New()
+// function and hopefully, some time in the future, we can get rid of the singleton entirely.
+type Encryption interface {
+	// ApplyEnvConfigurations applies a configuration coming from the operating system environment.
+	ApplyEnvConfigurations(
+		encryption map[encryptionconfig.Key]encryptionconfig.Config,
+		decryptionFallback map[encryptionconfig.Key]encryptionconfig.Config,
+	) error
+
+	// ApplyHCLEncryptionConfiguration applies a single encryption configuration coming from HCL.
+	ApplyHCLEncryptionConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error
+	// ApplyHCLDecryptionFallbackConfiguration applies a single decryption fallback configuration coming from HCL.
+	ApplyHCLDecryptionFallbackConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error
```

Comment:
I think it makes more sense for these methods to be independent of what's the source of configuration, and get it as a dependency. There's no actual difference in behaviour between these methods (which is great, as there shouldn't be).

WDYT about `ApplyConfigurations` + `ApplyEncryptionConfiguration` + `ApplyDecryptionFallbackConfiguration`, all of these getting the source as an argument? I think that this approach also makes at least one of these methods unnecessary (probably `ApplyConfigurations`, depending on usage)

Would appreciate to know what you think about this approach

Reply:
There is a fixed ruleset in what order these configurations need to be applied, see the description of the `Merge` function in `config.go`. If we remove the specialization in the API, we need to make the caller pass along the priority of that configuration, which would increase the overhead of the calling side.